Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -10,13 +10,97 @@ extension NSManagedObject {
     }
 }
 
+extension NSEntityDescription {
+    class func insertNewObject<T: NSFetchRequestResult>(
+        forEntityName entityName: String,
+        into context: NSManagedObjectContext,
+        forRequest request: NSFetchRequest<T>,
+        cachingInto cache: FetchCache
+    ) -> NSManagedObject {
+        let entity = insertNewObject(forEntityName: entityName, into: context)
+        cache.set(request, objectIds: [entity.objectID])
+        return entity
+    }
+}
+
+protocol NSFetchRequestGettable {}
+
+extension NSFetchRequestGettable where Self: NSManagedObject {
+    static func fetchRequest(id: String) -> NSFetchRequest<Self> {
+        let request = NSFetchRequest<Self>(entityName: entityName)
+        request.predicate = NSPredicate(format: "id == %@", id)
+        return request
+    }
+    
+    static func fetchRequest(keyPath: String, equalTo key: String) -> NSFetchRequest<Self> {
+        let request = NSFetchRequest<Self>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, key)
+        return request
+    }
+}
+
+extension NSManagedObject: NSFetchRequestGettable {}
+
 extension NSManagedObject {
+    static func load<T: NSManagedObject>(by id: String, context: NSManagedObjectContext) -> [T] {
+        load(keyPath: "id", equalTo: id, context: context)
+    }
+    
+    static func load<T: NSManagedObject>(keyPath: String, equalTo key: String, context: NSManagedObjectContext) -> [T] {
+        let request = NSFetchRequest<T>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, key)
+        return load(by: request, context: context)
+    }
+    
     static func load<T>(by request: NSFetchRequest<T>, context: NSManagedObjectContext) -> [T] {
+        request.entity = NSEntityDescription.entity(forEntityName: entityName, in: context)!
         do {
-            return try context.fetch(request)
+            return try context.fetch(
+                request,
+                using: FetchCache.shared
+            )
         } catch {
             log.error("Could not load \(error), \(error.localizedDescription)")
             return []
         }
     }
 }
+
+class FetchCache {
+    static let shared = FetchCache()
+    
+    private var cache = [NSFetchRequest<NSFetchRequestResult>: [NSManagedObjectID]]()
+    
+    func set<T>(_ request: NSFetchRequest<T>, objectIds: [NSManagedObjectID]) where T: NSFetchRequestResult {
+        cache[request as! NSFetchRequest<NSFetchRequestResult>] = objectIds
+    }
+    
+    func get<T>(_ request: NSFetchRequest<T>) -> [NSManagedObjectID]? where T: NSFetchRequestResult {
+        cache[request as! NSFetchRequest<NSFetchRequestResult>]
+    }
+    
+    func clear() {
+        cache.removeAll()
+    }
+}
+
+extension NSManagedObjectContext {
+    func fetch<T>(_ request: NSFetchRequest<T>, using cache: FetchCache) throws -> [T] where T: NSFetchRequestResult {
```

Comment:
This only takes into consideration using the same request. But it can also happen that objects created with a request are fetched individually in another place.

Also, the cache needs to be properly protected against deletions (not only a complete clear), which is not at the moment. (eg. I fetch the list of channels, then an event comes saying that a model has been deleted -> from that point onwards the cache is keeping a reference to an object that is deleted)
Same goes for updates (not 100% sure about this one)

Reply:
>This only takes into consideration using the same request. But it can also happen that objects created with a request are fetched individually in another place.

Caching objects based on their "id"s is tricky, certainly possible but I don't think it's something we need to investigate at this point

> (eg. I fetch the list of channels, then an event comes saying that a model has been deleted -> from that point onwards the cache is keeping a reference to an object that is deleted)

since cache is cleared between DB writes, it's not possible to fetch a channel that's been deleted from the cache. we fetch a list of channels, cache is cleared, `channel.delete` events comes, channel is cached, mutated, DB flush, cache cleared. if we fetch the list again, the channel is not in cache anymore, and if it's deleted from DB it won't be fetched to the cache

Reply:
It would be nice if there was some way of caching objects by their Stream ID (even though it's tricky). If you would go down that route, you can prewarm the cache by picking apart the JSON and, for example, do one fetch request for all UserDTO objects by user ids found in the JSON. 

Let's say that you have a JSON with 50 unique users, you could save lots of queries if those users already exist in the database because you can preload them all into the cache in one go.

Reply:
@donnywals that's good feedback for phase 2! Notion doc has some steps for that